### PR TITLE
IPFilter: Load and optimize rules in pairs

### PR DESCRIPTION
**Pull request by jonask84**
This patch aims to improve rule compiler optimization runtime.
It does so by changing the order rules are compiled into a program tree.

The original algorithm constructs the program tree by sequentially
appending each of the rules to the end of one single tree,
then running one single optimize operation on that whole tree.

Since the optimization algorithm has a polynomial runtime, this can get
prohibitively slow for large trees.

This patch loads rules two-by-two into small trees, optimizes them, then
combines these trees into a series of smaller, optimized trees. The process
is then repeated until there is only one tree left.

By optimizing each tree as it goes, the new algorithm keeps the tree
size as small as possible, avoiding the polynomial growth in runtime.

The reduction in rule load times varies with the rule set in question,
however we are seeing a reduction in load times of roughly 1/3 on
average. In some cases as much as 97% reduction.

The load time reduction varies quite a bit by rule set, but on average
we are seeing about 1/3 reduction. In some cases the reduction is as
great as 97%, going from 1150ms to 29ms (This rule set had 44 rules, and
45k tree nodes.) These numbers were obtained by running click testies
with various rule sets on server hardware.

### CODE REVIEW

**Comment by jonask84** [comment]
Hi there! Not sure I created this pull request correctly, but I hope we can at least get the evaluation process started. 

 Basically what this patch aims to do is to improve load time associated with rule-tree compilation and optimization.

I'll be happy to respond to any feedback or questions!

The very best,
- Jonas Ask

**Comment by rtshanks** [comment]
love it.  rocking the open source.

**Comment by tbarbette** [comment]
Nice ! I saw recently in the ClickNP paper (https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/main-4.pdf) that they took Click out of the "rule insertion time" graph as it was way too slow. I guess this partially solves it.

**Comment by kohler** [comment]
I am very excited about this, and it's a great idea. I'm going to create a branch on kohler/click and check in some nits first, and ask what you think. It would be LOVELY to see the 97%-reduction runtime.

File: elements/ip/ipfilter.cc [hunk 1]
```diff
@@ -1208,12 +1208,40 @@ IPFilter::Parser::parse_test(int pos, bool negated)
...
+    int step_offset = p1.ninsn();
+    int jump_target = step_offset;
+    if (p1.ninsn() == 0) {
+        // p1 program is empty.
+        // If it has a valid target, we always jump to it. p2 is ignored.
+        // If target is invalid, proceed as normal. Resulting tree == p2.
+        int target = -p1.output_everything();
+
+        //check if target is valid
+        if (target != Classification::j_never + 1) {
```
Comment (kohler) [inline elements/ip/ipfilter.cc]: I know what `Classification::j_never` means and I know what `Classification::j_never + 1` means (it's referenced in elements/standard/classification.cc). But this is `(-Classification::j_never) + 1`, which is an integer overflow. I'm going to write this to what I think you meant but a test is really required.

File: elements/ip/ipfilter.cc [hunk 2]
```diff
@@ -1208,12 +1208,40 @@ IPFilter::Parser::parse_test(int pos, bool negated)
...
+        jump_target = -p2.output_everything();
+    }
+
+    p1.redirect_unfinished_insn_tree(jump_target);
+    p2.offset_insn_tree(step_offset);
+
+    for (int i = 0; i < p2.ninsn(); i++) {
+       p1.add_raw_insn(p2.insn(i));
+    }
+}
```
Comment (kohler) [inline elements/ip/ipfilter.cc]: I think this misses the case that both programs have 0 insns, but the first goes to an invalid target & the second goes to a valid one

Comment (kohler) [inline elements/ip/ipfilter.cc]: Example IPFilter config: `IPFilter(0 false, 0 false, 0 false, 0 true)`

Comment (kohler) [inline elements/ip/ipfilter.cc]: This also should reset `p1._output_everything`. Example config: `IPFilter(0 false, 0 false, 0 ip src 10.0.0.1)`

File: elements/ip/ipfilter.cc [hunk 3]
```diff
@@ -1266,19 +1296,41 @@ IPFilter::parse_program(Classification::Wordwise::CompressedProgram &zprog,
...
+            }
+            n = n * 2;
+        }
+
+        //check if there is an odd program at the end not yet merged
+        if (n != num_progs) {
+            merge_programs(progs[0], progs[n]);
+            progs[0].optimize(offset_map, offset_map + 2, Classification::offset_max);
+        }
+    }
```
Comment (kohler) [inline elements/ip/ipfilter.cc]: fails to route final `j_failure` branches to `j_never`, which is required for configs like `IPFilter(0 ip src 10.0.0.1)`

**Comment by kohler** [comment]
OK! I added a test (IPFilter-07) for various edge cases, and also added a commit that cleans up the code as I think it should be (see 190f09191d2f85af0bd47fd50c562b1fd9269376). If you agree, I merge! THANKS!

**Comment by jonask84** [comment]
Hi guys!

Thank you so much for looking at the code!
This is very exciting, I'll do my best to reply to the feedback inline.

Cheers!
- Jonas

File: elements/ip/ipfilter.cc (hunk 1 above)
Comment (jonask84) [inline elements/ip/ipfilter.cc]: It's been a few months since I wrote this, so please forgive me if I misunderstood some of the concepts for forget something!

Yes, so, this particular code is for when merging two programs/trees, p1 and p2, where p1 is empty.

The way rules are evaluated sequentially means P1 decisions override p2, so if P1 is empty and points to a valid rule 'target', then the whole of p2 becomes irrelevant. 

This check exits the function early (thus dropping p2) if p1 has a valid target, which I've defined to be "anything that's not an invalid target".

From reading the code and experimenting, it seemed that invalid targets have the numerical value of "j_never + 1". 

The Jumps enum in classification.hh, sets j_never to -2147483647, and so that +1 would be -2147483646. Which is also the value of the 'j_failure' enum.

It could be that I misunderstood this part, but in debugging that seemed to be the value the jumps were set to when they were "unlinked" or invalid.

**Comment by jonask84** [comment]
The compressed code looks really great!

I'm going over it now (as best as I can; you know this stuff way better than I do), and it's looking very nice.

I did run into an issue though, that I'll look into deeper later tonight.
I was running my suite of edge cases, and I hit on a case that will trip the assertion on line 924 in classification.cc (`assert(failure <= 0);`)

It triggers on this rule set: 
`Idle -> c::IPFilter(allow false, allow false, allow false, allow false, allow src 10.0.0.1) -> Idle;`

I have to look into it closer, but for now I want to run my other edge cases too (I did make a testie for it, but it wasn't really ready for mainline :)

**Comment by kohler** [comment]
Very good catch! Added another commit to the kohler/click branch to catch this. Problem was that the signs of `_output_everything` and `insn` branches are opposite.

**Comment by jonask84** [comment]
That makes sense, when I was working my way through this (and the compiler/optimizer) back in February the jumps to rules or other nodes and their sign took a while for me to get used to!

This now passes any test I made, and the code looks really good. Added a comment on the actual commit for completeness. 

On a similar note, I think I found a rule set combination that causes the compiler to misbehave. In short it appears to optimize away a rule that it shouldn't. It might be that I've missed something, but I think it would be good if you had a look at it.

I have a testie (though based on the meraki framework) that exercises the issue. If you like to take a look at it, what is the best way to send it to you?

Cheers!
- Jonas

**Comment by kohler** [comment]
I'd love to look at it. Send to ekohler@gmail.com if you can't post it. But what you send I may add as a test.

**Comment by jonask84** [comment]
Awesome! There's nothing proprietary about it, so I created a new issue on the main branch.
My apologies if it turns out to be just an issue with our fork (though I don't believe that is the case, I don't think anyone's dared to make changes to the rules compiler)

**Comment by kohler** [comment]
Merged, thanks!
